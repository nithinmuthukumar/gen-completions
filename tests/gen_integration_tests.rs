//! Test generating completions from JSON files

use std::{
  env,
  path::{Path, PathBuf},
  process::{Command, Stdio},
};

use assert_cmd::prelude::{CommandCargoExt, OutputAssertExt};
use insta::Settings;

const BIN_NAME: &str = "gen-completions";

fn run_test(shell: &str, conf: &str, args: &[&str]) {
  // The project's root directory
  let root = env::var("CARGO_MANIFEST_DIR").unwrap();

  let conf_file = PathBuf::from(root).join("tests/resources/gen").join(conf);

  // The gen-completions binary to test
  let mut cmd = Command::cargo_bin(BIN_NAME).unwrap();
  let cmd = cmd.arg("for").arg(shell).arg(&conf_file).args(args);
  // So we can explicitly ask for logging
  if let Ok(log_level) = env::var("RUST_LOG") {
    cmd.env("RUST_LOG", log_level).stderr(Stdio::inherit());
  }
  let got = cmd.output().unwrap().stdout;
  let got = std::str::from_utf8(&got).unwrap().trim();
  cmd.assert().success();

  let mut settings = Settings::clone_current();
  settings.set_snapshot_path(Path::new("snapshots/gen/"));
  settings.set_snapshot_suffix(format!("{}.{}", conf, shell));
  settings.set_description(format!(
    "Generated for shell {} using config file {}",
    shell, conf
  ));
  settings.set_input_file(conf_file);
  settings.bind(|| {
    insta::assert_snapshot!(got);
  });
}

#[test]
fn test1_zsh() {
  run_test("zsh", "test1.json", &[]);
}

#[test]
fn test1_bash() {
  run_test("bash", "test1.json", &[]);
}

#[test]
fn test1_nu() {
  run_test("nu", "test1.json", &[]);
}

#[test]
fn test1_kdl() {
  run_test("kdl", "test1.json", &[]);
}

#[test]
fn test1_json() {
  run_test("json", "test1.json", &[]);
}

// #[test]
// fn types_bash() {
//   run_test("bash", "test-types.kdl", &[]);
// }

// #[test]
// fn types_zsh() {
//   run_test("zsh", "test-types.kdl", &[]);
// }

#[test]
fn types_nu() {
  run_test("nu", "test-types.kdl", &[]);
}
