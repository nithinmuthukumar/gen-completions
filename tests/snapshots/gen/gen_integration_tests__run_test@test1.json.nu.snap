---
source: tests/gen_integration_tests.rs
description: Generated for shell nu using config file test1.json
expression: got
input_file: tests/resources/gen/in/test1.json
---
export extern "test1" [
  --h(-h) # Show help information
  --verbose(-v) # Verbose output
  --loud # Verbose output
]

export extern "test1 sub1" [
  --foobar # Something something [brackets]
]

export extern "test1 sub1 nested" [
  --command # Run a command or something
  --another-invalid-flag?! # Run a command or something
  --install # Install a thing
]

export extern "test1 sub2" [
  --a # Both options should be picked up even though the short one is weird
  --all # Both options should be picked up even though the short one is weird
  --backupdir(-C) # The short form should be picked up as -C, not -Cdirectory (example from nano)
]
